tendermint/node.go

```diff
@@ -0,0 +1,75 @@
+package tendermint
+
+import (
+	"os"
+	"path/filepath"
+	"time"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/config"
+	"github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/node"
+	"github.com/tendermint/tendermint/p2p"
+	"github.com/tendermint/tendermint/privval"
+	"github.com/tendermint/tendermint/proxy"
+	"github.com/tendermint/tendermint/types"
+)
+
+// NewNode retruns new tendermint node that runs the app.
+func NewNode(logger log.Logger, app abci.Application, root, seeds, externalAddress string, validatorPubKey crypto.PubKey) (*node.Node, error) {
+	if err := os.MkdirAll(filepath.Join(root, "config"), 0755); err != nil {
+		return nil, err
+	}
+	if err := os.MkdirAll(filepath.Join(root, "data"), 0755); err != nil {
+		return nil, err
+	}
+
+	cfg := config.DefaultConfig()
+	cfg.P2P.PersistentPeers = seeds
+	// cfg.P2P.Seeds = seeds
+	// cfg.P2P.SeedMode = true
+	cfg.P2P.AddrBookStrict = false
+	cfg.P2P.AllowDuplicateIP = true
+	cfg.P2P.ExternalAddress = externalAddress
+	cfg.Consensus.TimeoutCommit = 10 * time.Second
+	cfg.SetRoot(root)
+
+	nodeKey, err := p2p.LoadOrGenNodeKey(cfg.NodeKeyFile())
+	if err != nil {
+		return nil, err
+	}
+
+	me := privval.LoadOrGenFilePV(cfg.PrivValidatorKeyFile(), cfg.PrivValidatorStateFile())
+
+	return node.NewNode(cfg,
+		me,
+		nodeKey,
+		proxy.NewLocalClientCreator(app),
+		genesisLoader(validatorPubKey),
+		node.DefaultDBProvider,
+		node.DefaultMetricsProvider(cfg.Instrumentation),
+		logger,
+	)
+}
+
+func genesisLoader(validator crypto.PubKey) func() (*types.GenesisDoc, error) {
+	return func() (*types.GenesisDoc, error) {
+		genesis := &types.GenesisDoc{
+			GenesisTime:     time.Date(2019, 8, 8, 0, 0, 0, 0, time.UTC),
+			ChainID:         "xxx",
+			ConsensusParams: types.DefaultConsensusParams(),
```

Comment:
using the default consensus here but a config with 10 sec. Don't we have any issue with that?

Reply:
ConsensusParams != config.Consensus

config.Consensus is stored in `config/config.toml`:

```toml
##### consensus configuration options #####
[consensus]

wal_file = "data/cs.wal/wal"

timeout_propose = "3s"
timeout_propose_delta = "500ms"
timeout_prevote = "1s"
timeout_prevote_delta = "500ms"
timeout_precommit = "1s"
timeout_precommit_delta = "500ms"
timeout_commit = "5s"
# ....
```

ConsensusParams is stored in `config/genesis.json`:

```json
  "consensus_params": {
    "block": {
      "max_bytes": "22020096",
      "max_gas": "-1",
      "time_iota_ms": "1000"
    },
    "evidence": {
      "max_age": "100000"
    },
    "validator": {
      "pub_key_types": [
        "ed25519"
      ]
    }
  },
```